### ISSUE TITLE
Allow passing multiple FeatureOptions to `required:`

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#3941]
Passing multiple `FeatureOption` objects will do an `AND` on them all, f.ex.:

```meson
gtk_x11_dep = dependency('gtk+-x11-3.0', version : '>= 3.10',
                         required : [get_option('x11'), get_option('gtk3')])
```
The table looks like:

| A | B | A & B |
|---|---|-------|
| enabled | enabled | enabled |
| enabled | disabled | disabled |
| enabled | auto | auto |
| disabled | disabled | disabled |
| disabled | auto | disabled |
| auto | auto | auto |

The only interesting case in this is when A is `enabled` and B is `auto`. I believe that the more useful behaviour in that case is to evaluate to `auto` for the dependency.

**Comment by ndufresne** [comment]
Would still be handy, here I got a gst case with examples and qt5 I would like to use.

**Comment by xclaesse** [comment]
Hmm, I remember I commented somewhere about this, but couldn't find it. I suggested adding methods on feature options, that would do something like: `feature = feature1.and(feature2)`. Which would allow defining other operations like `or()` with different logical table.

**Comment by nirbheek** [comment]
> I suggested adding methods on feature options, that would do something like: `feature = feature1.and(feature2)`. Which would allow defining other operations like `or()` with different logical table.

I've been wanting chaining of features, but wasn't sure of the semantics or syntax. I like these semantics a lot, but can we use the usual `and` and `or` operators instead?

**Comment by bonzini** [comment]
Makes sense even with the methods that are now available in Feature; but I would have naively expected to get A|B if an array is passed to `required: ...`. For example:

```
A           B           A|B
enabled 	enabled 	enabled
enabled 	disabled 	enabled
enabled 	auto 	    enabled
disabled 	disabled 	disabled
disabled 	auto 	    auto
auto 	    auto 	    auto
```

(and also true -> enabled, false -> auto). The idea is that you could do

```
openssl = dependency('openssl', required: [get_option('tls'), get_option('dtls')]
if get_option('tls')
  src += 'tls.c'
endif
if get_option('dtls')
  src += 'dtls.c'
endif
```

Maybe given the confusion this is not a "TOOWTDI"-ish idea after all...

**Comment by nirbheek** [comment]
I think defining the truth table for AND, OR, NOT operations on FeatureOptions is clearer, and allows you to chain options with other boolean values in the build file.

**Comment by dcbaker** [comment]
I had looked at this in the past, and my thought had been to add a module that could do that kind of work: https://github.com/mesonbuild/meson/pull/11885 was the start.

It kind bogged down because I also wanted a `new()` method to create features, but my thought had been to have something like: `features.all(Feature, Feature, Feature..., mode : FeatureMode) -> Feature`, `features.any(Feature, Feature, Feature.., mode : FeatureMode) -> Feature`, `features.xor(Feature, Feature) -> Feature` (all and any are basically just and and or with variadic arguments, but I think we'd need some kind of keyword to control the semantics of auto + !auto in the `all` and `any`. I think the XOR case is one where you expect to get one and exactly one enabled and the other !enabled.

My thought is that we want something like:
- `enabled`/`disabled` - find one of those explicitly
- `allowed`` - find anything that would match `features.allowed()`, i.e., `enabled` | `auto`
- `~allowed` - find anything that matches `disabled` | `auto`

(the names of these modes needs some work)

> I think defining the truth table for AND, OR, NOT operations on FeatureOptions is clearer, and allows you to chain options with other boolean values in the build file.

If I'm reading what you're saying correctly you mean to be able to do something like:
``` meson
x = feature.new('auto') and false
```

I am strongly opposed to that idea, that sounds like the sort of thing that seems great until you start seeing the real world hard to find and fix bugs, kinda like javascripts `==` type coercion, or Python's `str` isnstance `Sequence[str]` problem.

I do have a strong interest in solving this, as Mesa has a lot of feature option interactions, and this kind of work would clean a bunch of them up nicely.
